### PR TITLE
Bump mypy and fix compatibility issue with 0.920

### PR DESCRIPTION
**Pull request by Apakottur**
Bump mypy to 0.920, and fix compatibility issues with the new version.

## Details

Mypy introduced a breaking change in it's API in https://github.com/python/mypy/pull/9951, where `TypeVarDef` and `TypeVarType` were merged into one class.

In addition to bumping the mypy version, we fix the compatibility issue with a simple check which preserves support for older versions of mypy.

This fix is similar to the one in Pydantic - https://github.com/samuelcolvin/pydantic/pull/3175

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Yossi Rozantsev for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/utils/await_maybe.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 
 async def await_maybe(value: AwaitableOrValue):
-    if inspect.iscoroutine(value):
+    if inspect.isawaitable(value):
```
Comment (BryceBeagle) [inline strawberry/utils/await_maybe.py:11]: Is this change related to the version bump?

File: strawberry/ext/mypy_plugin.py [hunk 2]
```diff
@@ -507,6 +520,7 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
                 line=stmt.line,
                 column=stmt.column,
                 type=sym.type,
+                kw_only=False,
```
Comment (BryceBeagle) [inline strawberry/ext/mypy_plugin.py:523]: Is this change related to the version bump?

Comment (BryceBeagle) [reply strawberry/ext/mypy_plugin.py:523]: Unrelated to this PR, but why do we use the `dict` constructor here instead of using a `{}` literal, @patrick91?

File: RELEASE.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Release type: minor
```
Comment (patrick91) [inline RELEASE.md:1]:
```suggestion
Release type: patch
```

This should be a patch I think 😊

File: strawberry/ext/mypy_plugin.py (hunk 2 above)
Comment (patrick91) [reply strawberry/ext/mypy_plugin.py:523]: not 100% sure why this looks different than the mypy plugin: https://github.com/python/mypy/blob/master/mypy/plugins/dataclasses.py#L342

but if I did the change I used dict because I didn't want to change all the params to have quotes :D

File: strawberry/utils/await_maybe.py (hunk 1 above)
Comment (Apakottur) [reply strawberry/utils/await_maybe.py:11]: Yeah, mypy 0.910 is ok with this but mypy 0.920 gives:
`Incompatible types in "await" (actual type "CoroutineType", expected type "Awaitable[Any]")`.

We could alter the typing of the function to only accepts coroutines, instead of changing to `isawaitable`.

File: strawberry/ext/mypy_plugin.py (hunk 2 above)
Comment (Apakottur) [reply strawberry/ext/mypy_plugin.py:523]: Yeah the mypy dataclasses plugin added support for `kw_only` in https://github.com/python/mypy/pull/10867, so we need it here as well.
I just put `False` here to see if things work. I'll compare our plugin to theirs and match the implementation.

File: RELEASE.md (hunk 3 above)
Comment (Apakottur) [reply RELEASE.md]: Done.

File: strawberry/ext/mypy_plugin.py [hunk 4]
```diff
@@ -509,23 +558,22 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
...
-            try:
-                attribute = DataclassAttribute(**params)  # type: ignore
-            except TypeError:
+            # Support the addition of `info` in mypy 0.800 and `kw_only` in mypy 0.920
+            # without breaking backwards compatibility.
+            if _Version.VERSION >= Decimal("0.800"):
                 params["info"] = cls.info
-                attribute = DataclassAttribute(**params)  # type: ignore
+            if _Version.VERSION >= Decimal("0.920"):
+                params["kw_only"] = is_kw_only
```
Comment (Apakottur) [inline strawberry/ext/mypy_plugin.py:566]: @patrick91 @BryceBeagle 

Retaining backwards compatibility becomes difficult as a lot has been added to the dataclasses plugin in `0.920`, see all the functions below.
We can add guards like this all over the place but maybe it's better to just keep multiple files for different versions? Or drop support for old versions? What do you think?

File: strawberry/utils/await_maybe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Apakottur) [reply strawberry/utils/await_maybe.py:11]: I agree, but this line now becomes a mypy error, following the version bump.
See https://github.com/strawberry-graphql/strawberry/pull/1503#discussion_r771814580

Comment (patrick91) [reply strawberry/utils/await_maybe.py:11]: would you be come sending a separate PR just for this? 😊 

I think it is worth isolating this change 😊 I also want to make sure there's no breaking change (even though our tests are passing)

File: strawberry/ext/mypy_plugin.py (hunk 4 above)
Comment (patrick91) [reply strawberry/ext/mypy_plugin.py:566]: we don't support kwonly, so we could drop support for it entirely :)

File: strawberry/utils/await_maybe.py (hunk 1 above)
Comment (Apakottur) [reply strawberry/utils/await_maybe.py:11]: Sure, here's a separate PR - https://github.com/strawberry-graphql/strawberry/pull/1505 :)

File: strawberry/ext/mypy_plugin.py [hunk 5]
```diff
@@ -42,13 +42,19 @@
...
     UnionType,
     get_proper_type,
 )
 
 
+# Backwards compatible with the removal of `TypeVarDef` in mypy 0.920.
+try:
+    from mypy.types import TypeVarDef
+except ImportError:
+    TypeVarDef = TypeVarType
```
[the remark this answers is not in the document]
Comment (Apakottur) [reply strawberry/ext/mypy_plugin.py]: Done

File: strawberry/ext/mypy_plugin.py [hunk 6]
```diff
@@ -262,33 +274,42 @@ def is_dataclasses_field_or_strawberry_field(expr: Expression) -> bool:
...
     """
-
-    if is_dataclasses_field_or_strawberry_field(expr):
-        expr = cast(CallExpr, expr)
-
+    if (
+        isinstance(expr, CallExpr)
+        and isinstance(expr.callee, RefExpr)
+        and expr.callee.fullname == "dataclasses.field"
+    ):
```
Comment (patrick91) [inline strawberry/ext/mypy_plugin.py:288]: I think we should keep using `is_dataclasses_field_or_strawberry_field` otherwise it won't work for `strawberry.field` :)

Comment (Apakottur) [reply strawberry/ext/mypy_plugin.py]: I undid all the changes to the function, seems like they were just adding an error message which seems irrelevant.

File: strawberry/ext/mypy_plugin.py [hunk 7]
```diff
@@ -509,16 +531,14 @@ def collect_attributes(self) -> Optional[List[DataclassAttribute]]:
...
-            try:
-                attribute = DataclassAttribute(**params)  # type: ignore
-            except TypeError:
+            # Support the addition of `info` in mypy 0.800 and `kw_only` in mypy 0.920
+            # without breaking backwards compatibility.
+            if MypyVersion.VERSION >= Decimal("0.800"):
                 params["info"] = cls.info
-                attribute = DataclassAttribute(**params)  # type: ignore
+            if MypyVersion.VERSION >= Decimal("0.920"):
+                params["kw_only"] = False
```
Comment (patrick91) [inline strawberry/ext/mypy_plugin.py:539]: we'll need to change this after we merge: https://github.com/strawberry-graphql/strawberry/pull/1187

**Review by patrick91 — APPROVED** [review]
Looks great! thank you so much 😊

File: strawberry/ext/mypy_plugin.py (hunk 4 above)
Comment (Apakottur) [reply strawberry/ext/mypy_plugin.py]: Good point. I removed all the `kw_only` logic and just set it to `False` as before, if the version of mypy is >=0.920

**Comment by Apakottur** [comment]
I was able to run the plugin with mypy versions `0.920`, `0.910` and `0.800`. Should we try an older version of mypy in the tests, or is manual testing enough here?

**Comment by patrick91** [comment]
> I was able to run the plugin with mypy versions `0.920`, `0.910` and `0.800`. Should we try an older version of mypy in the tests, or is manual testing enough here?

should be fine I think, no need to make our ci more complicated 😊
